lib/github.ml

```diff
@@ -304,15 +304,67 @@ let check_tag ~dry_run vcs tag =
   if Vcs.tag_exists ~dry_run vcs tag then Ok ()
   else
     R.error_msgf
-      "CHANGES.md lists '%s' as the latest release, but no corresponding tag \
+      "CHANGES.md lists '%a' as the latest release, but no corresponding tag \
        has been found in the repository.@.Did you forget to call 'dune-release \
        tag' ?"
-      tag
+      Vcs.Tag.pp tag
 
 let assert_tag_exists ~dry_run tag =
   Vcs.get () >>= fun repo ->
   if Vcs.tag_exists ~dry_run repo tag then Ok ()
-  else R.error_msgf "%s is not a valid tag" tag
+  else R.error_msgf "%a is not a valid tag" Vcs.Tag.pp tag
+
+(* Resolve again in case of annotated tags (most common case).
+   This is a no-op for non-annotated tags. In case of error, we
+   can assume that the remote is different because we checked that we
+   have the tag locally. *)
+let determine_remote_tag_status ~local_rev ~remote_rev vcs =
+  match Vcs.commit_id ~commit_ish:remote_rev vcs with
+  | Ok resolved_rev when resolved_rev = local_rev && resolved_rev = remote_rev
+    ->
+      (* the resolved_rev was the same as remote, thus it is unannotated *)
```

Comment:
Isn't it the other way around?

If it's annotated, there's an intermediate object. `ls-remote` will return this object's hash which we need to resolve again locally to obtain the pointed commit hash.

Reply:
No, I think this is correct. If the `resolved_rev` is the same as the `remote_rev` that means they were pointing to the same commit, that is, there was no intermediate object.

Looking at the original source code on `main`:

https://github.com/ocamllabs/dune-release/blob/f3f38cc463eaa90d44661f8fa41c8eccc55ea2c7/lib/github.ml#L329-L341

`remote_rev_unpeeled` is what is `remote_rev` here and `remote_rev` is what this branch calls `resolved_rev`. The logic is indeed not great, so for the future I would suggest creating something like `Vcs.is_annotated_tag` but for I think this refactoring makes it at least better than what was before.